gateway/src/apicast/policy/headers.lua

```diff
@@ -0,0 +1,106 @@
+--- Headers policy
+-- This policy allows to include custom headers that will be sent to the
+-- upstream as well as modify or delete the ones included in the original
+-- request.
+-- Similarly, this policy also allows to add, modify, and delete the headers
+-- included in the response.
+
+local ipairs = ipairs
+
+local policy = require('apicast.policy')
+local _M = policy.new('Headers policy')
+
+local new = _M.new
+
+local function new_header_value(current_value, value)
+  if current_value then
+    return current_value .. ', ' .. value
+  else
+    return value
+  end
+end
+
+local function set_request_header(header, value)
+  ngx.req.clear_header(header)
+
+  if value and value ~= '' then
+    ngx.req.set_header(header, value)
+  end
+end
+
+local function add_request_header(header, value)
+  local current_value = ngx.req.get_headers()[header]
```

Comment:
We should check the performance of this. `get_headers` will read all the headers. That might not be needed.
You can access individual headers as ngx var `http_header_name` (`ngx.var.http_x_request_id` for example).

Reply:
Ok. Just did a little experiment.

For a request:

```
GET /
Foo: bar
Foo: foo
```

`ngx.req.get_headers()` returns: `foo = { "bar", "foo" }`
`ngx.var.http_foo` returns `"bar"`

That is very unfortunate.

If we need to use get_headers then it should be done once and not in a loop.

Reply:
From https://github.com/openresty/lua-nginx-module/issues/161#issuecomment-9116263 it looks like both the http_HEADER var and get_headers should have same performance, so better to use get_headers once.